opt/check_openmanage-3.7.12/man/check_openmanage.8
'\" t
.\"     Title: check_openmanage
.\"    Author: Trond Hasle Amundsen <t.h.amundsen@usit.uio.no>
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 07/28/2014
.\"    Manual: Nagios plugin
.\"    Source: check_openmanage
.\"  Language: English
.\"
.TH "CHECK_OPENMANAGE" "8" "07/28/2014" "check_openmanage" "Nagios plugin"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
check_openmanage \- Nagios plugin for checking the hardware status on Dell servers running OpenManage
.SH "SYNOPSIS"
.HP \w'\fBcheck_openmanage\fR\ 'u
\fBcheck_openmanage\fR [\fIoption\fR\&.\&.\&.]
.HP \w'\fBcheck_openmanage\fR\ 'u
\fBcheck_openmanage\fR \-H\ \fIhostname\fR [\fIoption\fR\&.\&.\&.]
.SH "DESCRIPTION"
.PP
\fBcheck_openmanage\fR
is a plugin for Nagios which checks the hardware health of Dell servers running OpenManage Server Administrator (OMSA)\&. The plugin checks the health of the storage subsystem, power supplies, memory modules, temperature probes etc\&., and gives an alert if any of the components are faulty or operate outside normal parameters\&.
.PP
\fBcheck_openmanage\fR
is designed to be used by either locally (using NRPE or similar) or remotely (using SNMP)\&. In either mode, the output is (nearly) the same\&. Note that checking the alert log is not supported in SNMP mode\&.
.SH "GENERAL OPTIONS"
.PP
\fB\-f\fR, \fB\-\-config \fR\fB\fIfile\fR\fR
.RS 4
Specify a configuration file\&. For reference on the config file syntax and options, consult the
\fBcheck_openmanage.conf\fR(5)
manual page\&.
.RE
.PP
\fB\-t\fR, \fB\-\-timeout \fR\fB\fIseconds\fR\fR
.RS 4
The number of seconds after which the plugin will abort\&. Default timeout is 30 seconds if the option is not present\&.
.RE
.PP
\fB\-p\fR, \fB\-\-perfdata\fR [\fIargument\fR]
.RS 4
Collect performance data\&. Performance data collected include temperatures (in Celsius) and fan speeds (in rpm)\&. On systems that support it, power consumption is also collected (in Watts)\&. This option takes one of two arguments, both of which are optional:
.PP
\fIminimal\fR
.RS 4
If
\fIminimal\fR
is specified as argument, the plugin will use shorter names for the performance data labels, e\&.g\&.
\(lqt0\(rq
instead of
\(lqtemp_0_system_board_ambient\(rq\&. This can be used as a workaround in cases where the plugin output needs shortening, for example if the 1024 character limit of NRPE is reached\&.
.RE
.PP
\fImultiline\fR
.RS 4
If
\fImultiline\fR
is specified as argument, the plugin will output the performance data on multiple lines, for Nagios 3\&.x and above\&.
.RE
.sp
The default behaviour should be sufficient for most users\&.
.RE
.PP
\fB\-\-legacy\-perfdata\fR
.RS 4
With version 3\&.7\&.0, performance data output changed\&. The new format is not compatible with the old format\&. Users who wish to postpone switching to the new performance data API may set this option\&.
.RE
.PP
\fB\-w\fR, \fB\-\-warning\fR \fIstring\fR | \fIfile\fR
.RS 4
Override the machine\-default temperature warning thresholds\&. Syntax is:
.sp
.if n \{\
.RS 4
.\}
.nf
\fIid1\fR=\fImax\fR[/\fImin\fR],\fIid2\fR=\fImax\fR[/\fImin\fR],\&.\&.\&.
	    
.fi
.if n \{\
.RE
.\}
.sp
The following example sets warning limits to max 50C for probe 0, and max 45C and min 10C for probe 1:
.sp
.if n \{\
.RS 4
.\}
.nf
check_openmanage \-w 0=50,1=45/10
	    
.fi
.if n \{\
.RE
.\}
.sp
The minimum limit can be omitted, if desired\&. Most often, you are only interested in setting the maximum thresholds\&.
.sp
This parameter can be either a string with the limits, or a file containing the limits string\&. The option can be specified multiple times\&.
.sp
\fINOTE\fR: This option should only be used to narrow the field of OK temperatures wrt\&. the OMSA defaults\&. To expand the field of OK temperatures, increase the OMSA thresholds\&. See the plugin web page for more information\&.
.RE
.PP
\fB\-c\fR, \fB\-\-critical\fR \fIstring\fR | \fIfile\fR
.RS 4
Override the machine\-default temperature critical thresholds\&. Syntax and behaviour is the same as for warning thresholds described above\&.
.RE
.PP
\fB\-F\fR, \fB\-\-fahrenheit\fR
.RS 4
Set Fahrenheit as unit for all temperatures\&. This option will override the
\fB\-\-tempunit\fR
option, if used simultaneously\&.
.RE
.PP
\fB\-\-tempunit\fR \fIunit\fR
.RS 4
Set temperature unit\&. Legal values are:
.sp
\fIF\fR: Fahrenheit
.sp
\fIC\fR: Celsius
.sp
\fIK\fR: Kelvin
.sp
\fIR\fR: Rankine
.sp
Default: C
.RE
.PP
\fB\-\-omreport\fR \fIpath\fR
.RS 4
Specify full path to omreport, if it is not installed in any of the regular places\&. Usually this option is only needed on Windows, if omreport is not installed on the C: drive\&.
.RE
.PP
\fB\-\-vdisk\-critical\fR
.RS 4
Make any alerts concerning virtual disks appear as critical\&.
.RE
.PP
\fB\-d\fR, \fB\-\-debug\fR
.RS 4
Debug output\&. Will report status on everything, even if status is ok\&. Blacklisted or unchecked components are ignored (i\&.e\&. no output)\&.
.sp
\fINOTE\fR: This option is intended for diagnostics and debugging purposes only\&. Do not use this option from within Nagios, i\&.e\&. in the Nagios config\&.
.RE
.PP
\fB\-h\fR, \fB\-\-help\fR
.RS 4
Display help message and exit\&.
.RE
.PP
\fB\-V\fR, \fB\-\-version\fR
.RS 4
Print version info and exit\&.
.RE
.SH "OUTPUT OPTIONS"
.PP
\fB\-o\fR, \fB\-\-ok\-info\fR \fIlevel\fR
.RS 4
This option lets you define how much output you want the plugin to give when everything is OK, i\&.e\&. the verbosity level\&. The default value is 0 (one line of output)\&. The output levels are cumulative\&.
.sp
\fI0\fR: Only one line
.sp
\fI1\fR: BIOS and firmware info on a separate line
.sp
\fI2\fR: Storage controller and enclosure info on separate lines
.sp
\fI3\fR: OMSA version on separate line
.sp
Default: 0
.sp
The reason that OMSA version is separated from the rest is that finding it requires running a really slow omreport command, when the plugin is run locally via NRPE\&.
.RE
.PP
\fB\-B\fR, \fB\-\-show\-blacklist\fR
.RS 4
If used together with blacklisting, this option will make the plugin output all blacklistings that are being used\&. The output will have the correct blacklisting syntax, and will make it easy to maintain control over which blacklistings that are used for each server, as any blacklistings can be viewed from Nagios\&.
.sp
When blacklisting is not used, this option has no effect\&.
.RE
.PP
\fB\-i\fR, \fB\-\-info\fR
.RS 4
Prefix any alerts with the service tag\&.
.RE
.PP
\fB\-e\fR, \fB\-\-extinfo\fR
.RS 4
Display a short summary of system information (model and service tag) in case of an alert\&.
.RE
.PP
\fB\-I\fR, \fB\-\-htmlinfo\fR [\fIcode\fR]
.RS 4
Using this option will make the servicetag and model name into clickable HTML links in the output\&. The model name link will point to the official Dell documentation for that model, while the servicetag link will point to a website containing support info for that particular server\&.
.sp
This option takes an optional argument, which should be a country or area code or\&. If the country code is omitted the servicetag link will still work, but it will not be speficic for your country or area\&. Example for Germany:
.sp
.if n \{\
.RS 4
.\}
.nf
check_openmanage \-\-htmlinfo de
	    
.fi
.if n \{\
.RE
.\}
.sp
If this option is used together with either the
\fB\-\-extinfo\fR
or
\fB\-\-info\fR
options, it is particularly useful\&. Only the most common country codes is supported at this time:
.TS
expand tab(:);
l s s
l l l
l l l
l l l
l l l
l l l
l l l
l l l
l s s
l s s
l l l
l l l
l s s
l s s
l l l
l l l.
T{
Europe, Middle East and Africa (EMEA)
T}
T{
\fIat\fR: Austria
T}:T{
\fIbe\fR: Belgium
T}:T{
\fIcz\fR: Czech Republic
T}
T{
\fIde\fR: Germany
T}:T{
\fIdk\fR: Denmark
T}:T{
\fIes\fR: Spain
T}
T{
\fIfi\fR: Finland
T}:T{
\fIfr\fR: France
T}:T{
\fIgr\fR: Greece
T}
T{
\fIit\fR: Italy
T}:T{
\fIil\fR: Israel
T}:T{
\fIme\fR: Middle East
T}
T{
\fIno\fR: Norway
T}:T{
\fInl\fR: The Netherlands
T}:T{
\fIpl\fR: Poland
T}
T{
\fIpt\fR: Portugal
T}:T{
\fIru\fR: Russia
T}:T{
\fIse\fR: Sweden
T}
T{
\fIuk\fR: United Kingdom
T}:T{
\fIza\fR: South Africa
T}:T{

T}
T{
\ \&
T}
T{
America
T}
T{
\fIbr\fR: Brazil
T}:T{
\fIca\fR: Canada
T}:T{
\fImx\fR: Mexico
T}
T{
\fIus\fR: USA
T}:T{

T}:T{

T}
T{
\ \&
T}
T{
Asia / Pacific
T}
T{
\fIau\fR: Australia
T}:T{
\fIcn\fR: China
T}:T{
\fIin\fR: India
T}
T{
\fIjp\fR: Japan
T}:T{

T}:T{

T}
.TE
.sp 1
.RE
.PP
\fB\-\-postmsg\fR \fIstring\fR | \fIfile\fR
.RS 4
User specified post message\&. Useful for displaying arbitrary or various system information at the end of alerts\&. The argument is either a string with the message, or a file containing that string\&. You can control the format with the following interpreted sequences:
.sp
\fI%m\fR: System model
.sp
\fI%s\fR: Service tag
.sp
\fI%b\fR: BIOS version
.sp
\fI%d\fR: BIOS release date
.sp
\fI%o\fR: Operating system name
.sp
\fI%r\fR: Operating system release
.sp
\fI%p\fR: Number of physical drives
.sp
\fI%l\fR: Number of logical drives
.sp

opt/check_openmanage-3.7.12/check_openmanage.php
<?php
#
# PNP4Nagios template for check_openmanage
#
# Author: 	Trond Hasle Amundsen
# Contact: 	t.h.amundsen@usit.uio.no
# Website:      http://folk.uio.no/trondham/software/check_openmanage.html
# Date: 	2014-07-28
# Version:      3.7.12
#
# Copyright (C) 2008-2012 Trond Hasle Amundsen
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# Array with different colors
$colors = array("0022ff", "22ff22", "ff0000", "00aaaa", "ff00ff",
		"ffa500", "cc0000", "0000cc", "0080C0", "8080C0",
		"FF0080", "800080", "688e23", "408080", "808000",
		"000000", "00FF00", "0080FF", "FF8000", "800000",
		"FB31FB");

# Counters
$f = 0;      # fan probe counter
$t = 0;      # temp probe counter
$a = 0;      # amp probe counter
$v = 0;      # volt probe counter
$e = 0;      # enclosure counter

# Flags
$visited_amp  = 0;

# IDs
$id_temp1 = 1;
$id_temp2 = 2;
$id_watt  = 3;
$id_amp   = 4;
$id_volt  = 5;
$id_fan   = 6;
$id_enc   = 7;

# Enclosure id
$enclosure_id = '';

# Default title
$def_title = 'Dell OpenManage';

# Temperature unit
if (!defined('tempunit_defined')) {
    define('tempunit_defined', 1);

    function tempunit($arg) 
    {
	$unit   = 'unknown';
	$vlabel = 'unknown';
	    
	switch ($arg) {
	default:
	    $vlabel = "Celsius";
	    $unit = "\xc2\xb0C";
	    break;
	case "F":
	    $vlabel = "Fahrenheit";
	    $unit = "\xc2\xb0F";
	    break;
	case "K":
	    $vlabel = "Kelvin";
	    $unit = "K";
	    break;
	case "R":
	    $vlabel = "Rankine";
	    $unit = "\xc2\xb0R";
	    break;
	}
	return array($unit, $vlabel);
    }
}


# Determine if we're running in legacy mode
$legacy = "no";
foreach ($this->DS as $KEY=>$VAL) {
    if (preg_match('/^(pwr_mon_|fan_|temp_|enclosure_|p\d+|f\d+|t\d+|e.+?t\d+)/', $VAL['LABEL'])) {
	$legacy = "yes";
	break;
    }
}

#------------------------------------------------------
#  MAIN LOOP
#------------------------------------------------------
if ($legacy == "yes") {  # --legacy--

    $count = 0;  # general counter
    $PWRcolor = "66FF00";

    # Flags
    $visited_fan  = 0;
    $visited_temp = 0;
    $visited_pwr  = 0;

    # Loop through the performance data
    foreach ($this->DS as $KEY=>$VAL) {
        
	# AMPERAGE PROBE
	if(preg_match('/^pwr_mon_/', $VAL['NAME']) || preg_match('/^p\d+[aw]$/', $VAL['NAME'])) {

	    # Watt
	    if (preg_match('/system/', $VAL['NAME']) || preg_match('/^p\d+w$/', $VAL['NAME'])) {
        
		# Long label
		$VAL['NAME'] = preg_replace('/^pwr_mon_\d+_/', '', $VAL['NAME']);
		$VAL['NAME'] = preg_replace('/_/', ' ', $VAL['NAME']);

		# Short label
		$VAL['NAME'] = preg_replace('/^p(\d+)w$/', 'Power Probe $1', $VAL['NAME']);

		++$count;
		$ds_name[$count] = "Power Consumption";
		$vlabel = "Watt";

		$title = $ds_name[$count];

		$opt[$count] = "--slope-mode --vertical-label \"$vlabel\" --title \"$def_title: $title\" ";

		if(isset($def[$count])){
		    $def[$count] .= rrd::def("var".$KEY, $VAL['RRDFILE'], $VAL['DS'], "AVERAGE");
		}
		else {
		    $def[$count] = rrd::def("var".$KEY, $VAL['RRDFILE'], $VAL['DS'], "AVERAGE");
		}
		$def[$count] .= rrd::area("var$KEY", "#".$PWRcolor, $VAL['NAME']);
		$def[$count] .= rrd::line1("var$KEY","#000000");
		$def[$count] .= rrd::gprint("var$KEY",array("LAST", "MAX", "AVERAGE"), "%6.0lf W");
	    }
	    # Ampere
	    elseif (preg_match('/current/', $VAL['NAME']) || preg_match('/^p\d+a$/', $VAL['NAME'])) {

		# Long label
		$VAL['NAME'] = preg_replace('/^pwr_mon_\d+_/', '', $VAL['NAME']);
		$VAL['NAME'] = preg_replace('/_/', ' ', $VAL['NAME']);
		$VAL['NAME'] = preg_replace('/ current \d+$/', '', $VAL['NAME']);
		$VAL['NAME'] = preg_replace('/ps/', 'PowerSupply', $VAL['NAME']);

		# Short label
		$VAL['NAME'] = preg_replace('/^p(\d+)a$/', 'Amperage Probe $1', $VAL['NAME']);
                
		if ($visited_pwr == 0) {
		    ++$count;
		    $visited_pwr = 1;
		}
		$ds_name[$count] = "Amperage Probes";
		$vlabel = "Ampere";

		$title = $ds_name[$count];

		$opt[$count] = "-X0 --lower-limit 0 --slope-mode --vertical-label \"$vlabel\" --title \"$def_title: $title\" ";
		if(isset($def[$count])){
		    $def[$count] .= rrd::def("var$KEY",$VAL['RRDFILE'],$VAL['DS'],"AVERAGE") ;
		}
		else {
		    $def[$count] = rrd::def("var$KEY",$VAL['RRDFILE'],$VAL['DS'],"AVERAGE") ;
		}
		$def[$count] .= rrd::line1("var$KEY", "#".$colors[$a], $VAL['NAME']) ;
		$def[$count] .= rrd::area("var$KEY","#".$colors[$a++]."20") ;
		$def[$count] .= rrd::gprint("var$KEY", array("LAST", "MAX", "AVERAGE"),"%4.2lf A") ;
	    }
	}

	# FANS (RPMs)
	if(preg_match('/^fan_/', $VAL['NAME']) || preg_match('/^f\d+$/', $VAL['NAME'])){
	    if ($visited_fan == 0) {
		++$count;
		$visited_fan = 1;
	    }

	    # Long label
	    $VAL['NAME'] = preg_replace('/^fan_\d+_/', '', $VAL['NAME']);
	    $VAL['NAME'] = preg_replace('/_rpm$/', '', $VAL['NAME']);
	    $VAL['NAME'] = preg_replace('/_/', ' ', $VAL['NAME']);

	    # Short label
	    $VAL['NAME'] = preg_replace('/^f(\d+)$/', 'Fan Probe $1', $VAL['NAME']);

	    $ds_name[$count] = "Fan Speed";

	    $opt[$count] = "-X0 --slope-mode --vertical-label \"RPMs\" --title \"$def_title: Fan Speeds\" ";
	    if(isset($def[$count])){
		$def[$count] .= rrd::def("var$KEY", $VAL['RRDFILE'], $VAL['DS'], "AVERAGE") ;
	    }
	    else {
		$def[$count] = rrd::def("var$KEY",$VAL['RRDFILE'],$VAL['DS'],"AVERAGE");
	    }
	    $def[$count] .= rrd::line1("var$KEY", "#".$colors[$f++],$VAL['NAME']);
	    $def[$count] .= rrd::gprint("var$KEY", array("LAST", "MAX", "AVERAGE"),"%6.0lf RPM");
	}
        
	# TEMPERATURES (Celsius)
	if(preg_match('/^temp_/', $VAL['NAME']) || preg_match('/^t\d+$/', $VAL['NAME'])){
	    if ($visited_temp == 0) {
		++$count;
		$visited_temp = 1;
	    }

	    # Long label
	    $VAL['NAME'] = preg_replace('/^temp_\d+_/', '', $VAL['NAME']);
	    $VAL['NAME'] = preg_replace('/_/', ' ', $VAL['NAME']);

	    # Short label
	    $VAL['NAME'] = preg_replace('/^t(\d+)$/', 'Temperature Probe $1', $VAL['NAME']);

	    $ds_name[$count] = "Chassis Temperatures";

	    $warnThresh = "INF";
	    $critThresh = "INF";

	    if ($VAL['WARN'] != "") {
		$warnThresh = $VAL['WARN'];
	    }
	    if ($VAL['CRIT'] != "") {
		$critThresh = $VAL['CRIT'];
	    }

	    $opt[$count] = "--slope-mode --vertical-label \"Celsius\" --title \"$def_title: Chassis Temperatures\" ";
	    if(isset($def[$count])){
		$def[$count] .= rrd::def("var$KEY",$VAL['RRDFILE'],$VAL['DS'],"AVERAGE") ;
	    }
	    else {
		$def[$count] = rrd::def("var$KEY", $VAL['RRDFILE'],$VAL['DS'],"AVERAGE") ;
	    }
	    $def[$count] .= rrd::line1("var$KEY", "#".$colors[$t++], $VAL['NAME']);
	    $def[$count] .= rrd::gprint("var$KEY",array("LAST", "MAX", "AVERAGE"),"%6.0lf C");
	}
        
	# ENCLOSURE TEMPERATURES (Celsius)
	if(preg_match('/^enclosure_(?P<id>.+?)_temp_\d+$/', $VAL['NAME'], $matches)
	   || preg_match('/^e(?P<id>.+?)t\d+$/', $VAL['NAME'], $matches)) {
	    $this_id = $matches['id'];

	    if ($enclosure_id != $this_id) {
		$e = 0;
		++$count;
		$enclosure_id = $this_id;
	    }

	    # Long label
	    $VAL['NAME'] = preg_replace('/^enclosure_.+?_temp_(\d+)$/', 'Probe $1', $VAL['NAME']);

	    # Short label
	    $VAL['NAME'] = preg_replace('/^e.+?t(\d+)$/', 'Probe $1', $VAL['NAME']);

	    $ds_name[$count] = "Enclosure $enclosure_id Temperatures";

	    $warnThresh = "INF";
	    $critThresh = "INF";

	    if ($VAL['WARN'] != "") {
		$warnThresh = $VAL['WARN'];
	    }
	    if ($VAL['CRIT'] != "") {
		$critThresh = $VAL['CRIT'];
	    }

	    $opt[$count] = "--slope-mode --vertical-label \"Celsius\" --title \"$def_title: Enclosure $enclosure_id Temperatures\" ";

	    if(isset($def[$count])){
		$def[$count] .= rrd::def("var$KEY", $VAL['RRDFILE'],$VAL['DS'],"AVERAGE") ;
	    }
	    else {
		$def[$count] = rrd::def("var$KEY",$VAL['RRDFILE'],$VAL['DS'],"AVERAGE") ;
	    }
	    $def[$count] .= rrd::line1("var$KEY","#".$colors[$e++],$VAL['NAME']) ;
	    $def[$count] .= rrd::gprint("var$KEY", array("LAST", "MAX", "AVERAGE"),"%6.0lf C");
	}
    }
}
else {  # --new--
	
    # Loop through the performance data
    foreach ($this->DS as $KEY=>$VAL) {
	
	$label = $VAL['LABEL'];

	# TEMPERATURES (AMBIENT)
	if (preg_match('/^T/', $label) && preg_match('/Ambient/', $label)) {

	    # Temperature unit and vertical label
	    list ($unit, $vlabel) = tempunit($VAL['UNIT']);

	    # Long label
	    $label = preg_replace('/^T(\d+)_(.+)/', '$2', $label);
	    $label = preg_replace('/_/', ' ', $label);

	    # Short label
	    $label = preg_replace('/^T(\d+)$/', 'Probe $1', $label);

	    $ds_name[$id_temp1] = "Temperatures";

	    $opt[$id_temp1] = "--slope-mode --vertical-label \"$vlabel\" --title \"$def_title: Ambient Temperature\" ";
	    if(isset($def[$id_temp1])){
		$def[$id_temp1] .= rrd::def("var$KEY",$VAL['RRDFILE'],$VAL['DS'],"AVERAGE") ;
	    }
	    else {
		$def[$id_temp1] = rrd::def("var$KEY", $VAL['RRDFILE'],$VAL['DS'],"AVERAGE") ;
	    }

	    # fancy graphing
	    $def[$id_temp1] .= rrd::gradient("var$KEY","#114480", "#1144dc", $label, 10, "50%");
	    $def[$id_temp1] .= rrd::gprint("var$KEY",array("LAST", "MAX", "AVERAGE"),"%5.1lf $unit");

	    # insert extra vertical space if we have thresholds
	    if ($VAL['WARN'] != "" || $VAL['CRIT'] != "") {
		$def[$id_temp1] .= "COMMENT:\\s ";
	    }

	    # warning threshold
	    if ($VAL['WARN'] != "") {
		$warnThresh = $VAL['WARN'];
		$def[$id_temp1] .= rrd::cdef("warn$KEY", "var$KEY,$warnThresh,GT,var$KEY,UNKN,IF");
		$def[$id_temp1] .= rrd::gradient("warn$KEY","#c4c400", "#ffff00","Above Upper Warning Threshold\: $warnThresh $unit\\l", 10, "50%");
	    }
	
	    # critical threshold
	    if ($VAL['CRIT'] != "") {
		$critThresh = $VAL['CRIT'];
		$def[$id_temp1] .= rrd::cdef("crit$KEY", "var$KEY,$critThresh,GT,var$KEY,UNKN,IF");
		$def[$id_temp1] .= rrd::gradient("crit$KEY","#800000", "#dc0000","Above Upper Critical Threshold\: $critThresh $unit\\l", 10, "50%");
	    }
	}

	# TEMPERATURES
	if (preg_match('/^T/', $label) && !preg_match('/Ambient/', $label)) {

	    # Temperature unit and vertical label
	    list ($unit, $vlabel) = tempunit($VAL['UNIT']);

	    # Long label
	    $label = preg_replace('/^T(\d+)_(.+)/', '$2', $label);
	    $label = preg_replace('/_/', ' ', $label);

	    # Short label
	    $label = preg_replace('/^T(\d+)$/', 'Probe $1', $label);

	    $ds_name[$id_temp2] = "Temperatures";

	    $opt[$id_temp2] = "--slope-mode --vertical-label \"$vlabel\" --title \"$def_title: Chassis Temperatures\" ";
	    if (isset($def[$id_temp2])) {
		$def[$id_temp2] .= rrd::def("var$KEY", $VAL['RRDFILE'],$VAL['DS'],"AVERAGE") ;
	    }
	    else {
		$def[$id_temp2] = rrd::def("var$KEY",$VAL['RRDFILE'],$VAL['DS'],"AVERAGE") ;
	    }
	    $def[$id_temp2] .= rrd::line1("var$KEY", "#".$colors[$t++], rrd::cut($label,20) );
	    $def[$id_temp2] .= rrd::gprint("var$KEY", array("LAST", "MAX", "AVERAGE"), "%4.1lf $unit");
	}

	# WATTAGE PROBE
	if (preg_match('/^W/', $label)) {

	    # Long label
	    $label = preg_replace('/^W(\d+)_(.+)/', '$2', $label);
	    $label = preg_replace('/_/', ' ', $label);

	    # Short label
	    $label = preg_replace('/^W(\d+)$/', 'Probe $1', $label);

	    $ds_name[$id_watt] = "Power Consumption";
	    $vlabel = "Watt";

	    $title = $ds_name[$id_watt];

	    $opt[$id_watt] = "--slope-mode --vertical-label \"$vlabel\" --title \"$def_title: $title\" ";

	    if(isset($def[$id_watt])){
		$def[$id_watt] .= rrd::def("var$KEY",$VAL['RRDFILE'], $VAL['DS'], "AVERAGE") ;
	    }
	    else {
		$def[$id_watt] = rrd::def("var$KEY",$VAL['RRDFILE'],$VAL['DS'],"AVERAGE") ;
	    }

	    # calculate kWh and BTU
            $def[$id_watt] .= "VDEF:tot$KEY=var$KEY,TOTAL ";
	    # rrd:vdef is broken in pnp4nagios 0.6.13 # $def[$id_watt] .= rrd::vdef("tot$KEY","var$KEY,TOTAL");
	    $def[$id_watt] .= rrd::cdef("kwh$KEY","var$KEY,POP,tot$KEY,1000,/,60,/,60,/");
	    $def[$id_watt] .= rrd::cdef("btu$KEY","kwh$KEY,3412.3,*");

	    # fancy graphing
	    $def[$id_watt] .= rrd::gradient("var$KEY","#800000","#dc0000","$label", 10, "50%");

	    # print avg, max and min
	    $def[$id_watt] .= rrd::gprint("var$KEY", array("LAST", "MAX", "AVERAGE"),"%6.0lf W");

	    # print kWh and BTU for time period
	    $def[$id_watt] .= rrd::comment("\\s");
	    $def[$id_watt] .= rrd::comment("    Total power used in time period\:");
	    $def[$id_watt] .= rrd::gprint("kwh$KEY","AVERAGE","%10.2lf kWh\l");
	    $def[$id_watt] .= rrd::comment("                                    ");
	    $def[$id_watt] .= rrd::gprint("btu$KEY","AVERAGE","%10.2lf BTU\l");
	}

	# AMPERAGE PROBE
	if (preg_match('/^A/', $label) && preg_match('/Current/', $label)) {

	    $first = 0;
	    if ($visited_amp == 0) {
		$first = 1;
		$visited_amp = 1;
	    }

	    # Long label
	    $label = preg_replace('/^A(\d+)_(.+)/', '$2', $label);
	    $label = preg_replace('/_/', ' ', $label);

	    # Short label
	    $label = preg_replace('/^A(\d+)$/', 'Probe $1', $label);

	    $ds_name[$id_amp] = "Amperage Probes";
	    $vlabel = "Ampere";

	    $title = $ds_name[$id_amp];

	    $opt[$id_amp] = "-X0 --lower-limit 0 --slope-mode --vertical-label \"$vlabel\" --title \"$def_title: $title\" ";
	    if(isset($def[$id_amp])){
		$def[$id_amp] .= rrd::def("var$KEY",$VAL['RRDFILE'], $VAL['DS'], "AVERAGE");
	    }
	    else {
		$def[$id_amp]  = rrd::def("var$KEY",$VAL['RRDFILE'], $VAL['DS'], "AVERAGE");
	    }

	    $space = strlen($label) < 16 ? str_repeat(' ', 16 - strlen($label)) : ' ';

	    $def[$id_amp] .= rrd::cdef("tier$KEY", "var$KEY,10,/");
	    $def[$id_amp] .= "AREA:tier$KEY#".$colors[$a]."b7::STACK ";
	    $def[$id_amp] .= "AREA:tier$KEY#".$colors[$a]."bf::STACK ";
	    $def[$id_amp] .= "AREA:tier$KEY#".$colors[$a]."c7::STACK ";
	    $def[$id_amp] .= "AREA:tier$KEY#".$colors[$a]."cf::STACK ";
	    $def[$id_amp] .= "AREA:tier$KEY#".$colors[$a]."d7::STACK ";
	    $def[$id_amp] .= "AREA:tier$KEY#".$colors[$a]."df::STACK ";
	    $def[$id_amp] .= "AREA:tier$KEY#".$colors[$a]."e7::STACK ";
	    $def[$id_amp] .= "AREA:tier$KEY#".$colors[$a]."ef::STACK ";
	    $def[$id_amp] .= "AREA:tier$KEY#".$colors[$a]."f7::STACK ";
	    $def[$id_amp] .= "AREA:tier$KEY#".$colors[$a]."ff:\"$label$space\":STACK ";
	    $a++;

	    if ($first) {
		$def[$id_amp] .= rrd::cdef("sum$KEY", "var$KEY,0,+");
	    }
	    else {
		$def[$id_amp] .= rrd::cdef("sum$KEY", "sum".($KEY-1).",var$KEY,+");
	    }
	    $def[$id_amp] .= rrd::line1("sum$KEY", "#555555");
	    $def[$id_amp] .= rrd::gprint("var$KEY", array("LAST", "MAX", "AVERAGE"),"%6.1lf A") ;
	}

	# VOLTAGE PROBE
	if (preg_match('/^V/', $label)) {

	    # Long label
	    $label = preg_replace('/^V(\d+)_(.+)/', '$2', $label);
	    $label = preg_replace('/_/', ' ', $label);

	    # Short label
	    $label = preg_replace('/^V(\d+)$/', 'Probe $1', $label);
		
	    $ds_name[$id_volt] = "Voltage Probes";
	    $vlabel = "Volts";

	    $title = $ds_name[$id_volt];

	    $opt[$id_volt] = "--slope-mode --vertical-label \"$vlabel\" --title \"$def_title: $title\" ";
	    if(isset($def[$id_volt])){
		$def[$id_volt] .= rrd::def("var$KEY", $VAL['RRDFILE'], $VAL['DS'],"AVERAGE");
	    }
	    else {
		$def[$id_volt] = rrd::def("var$KEY",$VAL['RRDFILE'],$VAL['DS'],"AVERAGE");
	    }
	    $def[$id_volt] .= rrd::line1("var$KEY", "#".$colors[$v++], rrd::cut($label,18) ) ;
	    $def[$id_volt] .= rrd::gprint("var$KEY", array("LAST", "MAX", "AVERAGE"),"%8.2lf V");
	}

	# FANS (RPMs)
	if (preg_match('/^F/', $label)) {

	    # Long label
	    $label = preg_replace('/^F(\d+)_(.+)/', '$2', $label);
	    $label = preg_replace('/_/', ' ', $label);

	    # Short label
	    $label = preg_replace('/^F(\d+)$/', 'Probe $1', $label);

	    $ds_name[$id_fan] = "Fan Probes";

	    $opt[$id_fan] = "-X0 --slope-mode --vertical-label \"RPMs\" --title \"$def_title: Fan Speeds\" ";
	    if(isset($def[$id_fan])){
		$def[$id_fan] .= rrd::def("var$KEY",$VAL['RRDFILE'],$VAL['DS'], "AVERAGE") ;
	    }
	    else {
		$def[$id_fan] = rrd::def("var$KEY",$VAL['RRDFILE'],$VAL['DS'], "AVERAGE") ; 
	    }
	    $def[$id_fan] .= rrd::line1("var$KEY", "#".$colors[$f++], rrd::cut($label,18) ) ;
	    $def[$id_fan] .= rrd::gprint("var$KEY", array("LAST", "MAX", "AVERAGE"), "%6.0lf RPM");
	}
	
	# ENCLOSURE TEMPERATURES (Celsius)
	if (preg_match('/^E(?P<encl>.+?)_t(emp_)?(?P<probe>\d+)/', $label, $matches)) {

	    # Temperature unit and vertical label
	    list ($unit, $vlabel) = tempunit($VAL['UNIT']);

	    $this_id     = $matches['encl'];
	    $probe_index = $matches['probe'];

	    if ($enclosure_id != $this_id) {
		$e = 0;
		$id_enc++;
		$enclosure_id = $this_id;
	    }

	    # Label
	    $label = "Probe $probe_index";

	    $ds_name[$id_enc] = "Enclosure $enclosure_id Temperatures";

	    $opt[$id_enc] = "--slope-mode --vertical-label \"$vlabel\" --title \"$def_title: Enclosure $enclosure_id Temperatures\" ";

	    if(isset($def[$id_enc])){
		$def[$id_enc] .= rrd::def("var$KEY", $VAL['RRDFILE'],$VAL['DS'], "AVERAGE") ;
	    }
	    else {
		$def[$id_enc] = rrd::def("var$KEY", $VAL['RRDFILE'],$VAL['DS'], "AVERAGE") ;
	    }
	    $def[$id_enc] .= rrd::line1("var$KEY", "#".$colors[$e++], rrd::cut($label, 14) );
	    $def[$id_enc] .= rrd::gprint("var$KEY", array("LAST", "MAX", "AVERAGE"),"%6.1lf $unit");
	}
    }
}

?>
